### ISSUE TITLE
Longest match first rule

### ISSUE THREAD

**Issue by kstaniek** [jpmens/mqttwarn#114]
I would propose implementing the "longest match first" rule in matching topics.
It allows to define more specific topic patters to be redirected to the different targets.
For example

    [/topic/#]
    targets = plugin:generic_target

    [/topic/specific]
    targets = plugin:specific_target

Then messages with topics /topic/a;  /topic/b, etc will go to generic_target 
and topic /topic/specific will go ONLY to specific_target but NOT to generic_target

The longer topic must take precedence over shorter and specific topic name over wildcard.
Please let me know if this makes any sense and I can start implementing it.
It can break the existing configuration as today message is replicated to ALL targets matching the configured topic patterns with paho.topic_matches_sub.
https://github.com/jpmens/mqttwarn/labels

**Comment by jpmens** [comment]
You're going to have to do some convincing here, because this breaks the notion of MQTT topic branches. In your example, "topic/#' would normally catch "topic/specific". I do understand what you're saying, but I wouldn't want to confuse people with 'short' vs. 'long' topic branch names, when MQTT has wildcards.

**Comment by kstaniek** [comment]
I'm having just a simple problem to be solved. Have majority of the MQTT topics matching /topic/# but for some specific topics I need a different target. Currently the request is being send to both. Maybe the solution for me will be using filer function. I feel that what I'm proposing may break a lot current deployments.

**Comment by jpmens** [comment]
It will break a lot, and I won't take this idea.

What I would entertain, if you're willing to provide code & docs, is the idea of having a new keyword which takes a function to _then_ do the dispatching. Taking your example:

```ini
[topic/#]
dispatcher = { "topic/#" : "plugin:generic_target", "topic/specific" : "plugin:specific_target" }
```

That is supposed to be an object containing topic `=>` plugin associations. I'm not sure this is really useful though, and it is probably not quite trivial.

**Comment by kstaniek** [comment]
What if instead of adding 'dispatcher' option I would extend the 'targets' option:
- If targets is a list then no change, works as usual.
- If targets is a dictionary then handled as proposed above.
I've got a code ready. Just wanted to to know what's your opinion.
My current implementation is that if there is a 'dispatcher' option then 'targets' option is ignored. But I'm not convinced this is clear. Adding new option needs to handle the situation when both options are used.
When extending the 'targets' options both actions are mutually exclusive by nature.

**Comment by kstaniek** [comment]
Please have a look at https://github.com/kstaniek/mqttwarn/blob/dispatcher/mqttwarn.py
I extended 'send_to_targets' and slightly corrected 'g' function in Config object to handle the SyntaxError exception. BTW: IMHO getdict, getlist and g functions must be reviewed.
If you agree, I will document the change and create a PR.

**Comment by kstaniek** [comment]
And documentation: https://github.com/kstaniek/mqttwarn/blob/dispatcher/README.md
See the The [__topic__] sections.
Please let me know if you like it, then will create PR.

**Comment by jpmens** [comment]
I've glanced at the code and have read the README. Both look very good to me, and I'll take it (because you've documented this very well!). Please submit a PR, and thank you.

**Comment by jpmens** [comment]
Closed via #115
